File: frontend/src/app/services/vehicle/vehicle.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { VehicleBrandDTO } from '../../dto/vehicle/VehicleBrandDTO';
import { VehicleModelDTO } from '../../dto/vehicle/VehicleModelDTO';
import { VehicleResponseDTO } from '../../dto/vehicle/VehicleResponseDTO';
import { CreateVehicleDTO } from '../../dto/vehicle/CreateVehicleDTO';
import { UpdateVehicleDTO } from '../../dto/vehicle/UpdateVehicleDTO';
import { VehicleLocationDTO } from '../../dto/vehicle/VehicleLocationDTO';
import { DistanceStatisticsDTO } from '../../dto/vehicle/DistanceStatisticsDTO';
import { AvailabilityStatisticsDTO } from '../../dto/vehicle/AvailabilityStatisticsDTO';
import { PageResponseDTO } from '../../dto/common/PageResponseDTO';

@Injectable({
  providedIn: 'root'
})
export class VehicleService {
  private readonly baseUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  getAllBrands(): Observable<VehicleBrandDTO[]> {
    return this.http.get<VehicleBrandDTO[]>(`${this.baseUrl}vehicles/brands`);
  }

  getModelsByBrand(brandId: number): Observable<VehicleModelDTO[]> {
    return this.http.get<VehicleModelDTO[]>(`${this.baseUrl}vehicles/brands/${brandId}/models`);
  }

  getAll(): Observable<VehicleResponseDTO[]> {
    return this.http.get<VehicleResponseDTO[]>(`${this.baseUrl}vehicles`);
  }

  getAllPaged(page: number = 0, size: number = 20, sortBy: string = 'id', sortDir: string = 'asc'): Observable<PageResponseDTO<VehicleResponseDTO>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);
    return this.http.get<PageResponseDTO<VehicleResponseDTO>>(`${this.baseUrl}vehicles/paged`, { params });
  }

  getById(id: number): Observable<VehicleResponseDTO> {
    return this.http.get<VehicleResponseDTO>(`${this.baseUrl}vehicles/${id}`);
  }

  search(query: string): Observable<VehicleResponseDTO[]> {
    return this.http.get<VehicleResponseDTO[]>(`${this.baseUrl}vehicles/search`, {
      params: { query }
    });
  }

  searchPaged(query: string, page: number = 0, size: number = 20): Observable<PageResponseDTO<VehicleResponseDTO>> {
    const params = new HttpParams()
      .set('query', query)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponseDTO<VehicleResponseDTO>>(`${this.baseUrl}vehicles/search/paged`, { params });
  }

  create(data: CreateVehicleDTO, images: File[]): Observable<VehicleResponseDTO> {
    const formData = new FormData();
    formData.append('data', new Blob([JSON.stringify(data)], { type: 'application/json' }));
    images.forEach((image) => {
      formData.append('images', image, image.name);
    });
    return this.http.post<VehicleResponseDTO>(`${this.baseUrl}vehicles`, formData);
  }

  update(id: number, data: UpdateVehicleDTO, newImages?: File[]): Observable<VehicleResponseDTO> {
    const formData = new FormData();
    formData.append('data', new Blob([JSON.stringify(data)], { type: 'application/json' }));
    if (newImages && newImages.length > 0) {
      newImages.forEach((image) => {
        formData.append('images', image, image.name);
      });
    }
    return this.http.put<VehicleResponseDTO>(`${this.baseUrl}vehicles/${id}`, formData);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}vehicles/${id}`);
  }

  getVehicleImages(vehicleId: number): Observable<{
    images: { id: number; originalName: string; url: string }[];
  }> {
    return this.http.get<{
      images: { id: number; originalName: string; url: string }[];
    }>(`${this.baseUrl}files/vehicle/${vehicleId}/images`);
  }

  getLastLocation(vehicleId: number): Observable<VehicleLocationDTO> {
    return this.http.get<VehicleLocationDTO>(`${this.baseUrl}vehicles/${vehicleId}/location`);
  }

  getDistanceStatsByPeriod(vehicleId: number, period: string): Observable<DistanceStatisticsDTO> {
    const params = new HttpParams().set('period', period);
    return this.http.get<DistanceStatisticsDTO>(`${this.baseUrl}vehicles/${vehicleId}/distance/stats`, { params });
  }

  getDistanceStatsByDateRange(vehicleId: number, startDate: string, endDate: string): Observable<DistanceStatisticsDTO> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);
    return this.http.get<DistanceStatisticsDTO>(`${this.baseUrl}vehicles/${vehicleId}/distance/stats`, { params });
  }

  getAvailabilityStatsByPeriod(vehicleId: number, period: string): Observable<AvailabilityStatisticsDTO> {
    const params = new HttpParams().set('period', period);
    return this.http.get<AvailabilityStatisticsDTO>(`${this.baseUrl}vehicles/${vehicleId}/availability/stats`, { params });
  }

  getAvailabilityStatsByDateRange(vehicleId: number, startTime: string, endTime: string): Observable<AvailabilityStatisticsDTO> {
    const params = new HttpParams()
      .set('startTime', startTime)
      .set('endTime', endTime);
    return this.http.get<AvailabilityStatisticsDTO>(`${this.baseUrl}vehicles/${vehicleId}/availability/stats`, { params });
  }
}

File: backend/src/main/java/nvt/backend/services/vehicle/VehicleService.java
package nvt.backend.services.vehicle;

import lombok.RequiredArgsConstructor;
import nvt.backend.dto.common.PageResponseDTO;
import nvt.backend.dto.vehicle.AvailabilityStatisticsDTO;
import nvt.backend.dto.vehicle.CreateVehicleDTO;
import nvt.backend.dto.vehicle.DistanceStatisticsDTO;
import nvt.backend.dto.vehicle.UpdateVehicleDTO;
import nvt.backend.dto.vehicle.VehicleLocationDTO;
import nvt.backend.dto.vehicle.VehicleResponseDTO;
import nvt.backend.model.vehicle.Vehicle;
import nvt.backend.model.vehicle.VehicleBrand;
import nvt.backend.model.vehicle.VehicleImage;
import nvt.backend.model.vehicle.VehicleModel;
import nvt.backend.repositories.vehicle.VehicleBrandRepository;
import nvt.backend.repositories.vehicle.VehicleLocationRepository;
import nvt.backend.repositories.vehicle.VehicleModelRepository;
import nvt.backend.repositories.vehicle.VehicleRepository;
import nvt.backend.services.storage.MinioService;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class VehicleService {

    private final VehicleRepository vehicleRepository;
    private final VehicleBrandRepository brandRepository;
    private final VehicleModelRepository modelRepository;
    private final VehicleLocationRepository locationRepository;
    private final VehicleTelemetryService telemetryService;
    private final MinioService minioService;

    @Value("${minio.bucket.vehicle-images}")
    private String vehicleImagesBucket;

    @Transactional
    @Caching(evict = {
        @CacheEvict(value = "vehiclesPage", allEntries = true),
        @CacheEvict(value = "vehicleSearch", allEntries = true)
    })
    public VehicleResponseDTO create(CreateVehicleDTO dto, List<MultipartFile> images) throws IOException {
        if (vehicleRepository.existsByLicensePlate(dto.getLicensePlate())) {
            throw new RuntimeException("Vehicle with this license plate already exists");
        }

        VehicleBrand brand = brandRepository.findById(dto.getBrandId())
                .orElseThrow(() -> new RuntimeException("Brand not found"));

        VehicleModel model = modelRepository.findById(dto.getModelId())
                .orElseThrow(() -> new RuntimeException("Model not found"));

        if (!model.getBrand().getId().equals(brand.getId())) {
            throw new RuntimeException("Model does not belong to the selected brand");
        }

        Vehicle vehicle = new Vehicle();
        vehicle.setLicensePlate(dto.getLicensePlate());
        vehicle.setWeightLimit(dto.getWeightLimit());
        vehicle.setBrand(brand);
        vehicle.setModel(model);

        vehicle = vehicleRepository.save(vehicle);

        minioService.createBucketIfNotExists(vehicleImagesBucket);

        for (MultipartFile file : images) {
            String minioPath = minioService.uploadFile(file, vehicleImagesBucket, "vehicle-" + vehicle.getId());

            VehicleImage img = new VehicleImage();
            img.setOriginalName(file.getOriginalFilename());
            img.setMinioPath(minioPath);
            img.setMinioBucket(vehicleImagesBucket);
            img.setVehicle(vehicle);
            vehicle.getImages().add(img);
        }

        vehicle = vehicleRepository.save(vehicle);

        return VehicleResponseDTO.fromEntity(vehicle);
    }

    public List<VehicleResponseDTO> getAll() {
        return vehicleRepository.findAllWithDetails().stream()
                .map(VehicleResponseDTO::fromEntity)
                .toList();
    }

    @Cacheable(value = "vehiclesPage", key = "#page + '-' + #size + '-' + #sortBy + '-' + #sortDir")
    public PageResponseDTO<VehicleResponseDTO> getAllPaged(int page, int size, String sortBy, String sortDir) {
        Sort sort = sortDir.equalsIgnoreCase("desc") ? Sort.by(sortBy).descending() : Sort.by(sortBy).ascending();
        Pageable pageable = PageRequest.of(page, size, sort);
        
        Page<Long> idPage = vehicleRepository.findAllIds(pageable);
        List<Vehicle> vehicles = idPage.getContent().isEmpty() 
                ? List.of() 
                : vehicleRepository.findAllWithDetailsByIds(idPage.getContent());
        
        return PageResponseDTO.<VehicleResponseDTO>builder()
                .content(vehicles.stream().map(VehicleResponseDTO::fromEntity).toList())
                .page(idPage.getNumber())
                .size(idPage.getSize())
                .totalElements(idPage.getTotalElements())
                .totalPages(idPage.getTotalPages())
                .first(idPage.isFirst())
                .last(idPage.isLast())
                .build();
    }

    @Cacheable(value = "vehicleById", key = "#id")
    public VehicleResponseDTO getById(Long id) {
        Vehicle vehicle = vehicleRepository.findByIdWithDetails(id)
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));
        return VehicleResponseDTO.fromEntity(vehicle);
    }

    public List<VehicleResponseDTO> search(String query) {
        return vehicleRepository.searchVehicles(query).stream()
                .map(VehicleResponseDTO::fromEntity)
                .toList();
    }

    @Cacheable(value = "vehicleSearch", key = "#query + '-' + #page + '-' + #size")
    public PageResponseDTO<VehicleResponseDTO> searchPaged(String query, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        
        Page<Long> idPage = vehicleRepository.searchVehicleIds(query, pageable);
        List<Vehicle> vehicles = idPage.getContent().isEmpty() 
                ? List.of() 
                : vehicleRepository.findAllWithDetailsByIds(idPage.getContent());
        
        return PageResponseDTO.<VehicleResponseDTO>builder()
                .content(vehicles.stream().map(VehicleResponseDTO::fromEntity).toList())
                .page(idPage.getNumber())
                .size(idPage.getSize())
                .totalElements(idPage.getTotalElements())
                .totalPages(idPage.getTotalPages())
                .first(idPage.isFirst())
                .last(idPage.isLast())
                .build();
    }

    @Transactional
    @Caching(evict = {
        @CacheEvict(value = "vehicleById", key = "#id"),
        @CacheEvict(value = "vehiclesPage", allEntries = true),
        @CacheEvict(value = "vehicleSearch", allEntries = true)
    })
    public VehicleResponseDTO update(Long id, UpdateVehicleDTO dto, List<MultipartFile> newImages) throws IOException {
        Vehicle vehicle = vehicleRepository.findByIdWithDetails(id)
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));

        if (dto.getLicensePlate() != null && !dto.getLicensePlate().equals(vehicle.getLicensePlate())) {
            if (vehicleRepository.existsByLicensePlateAndIdNot(dto.getLicensePlate(), id)) {
                throw new RuntimeException("Vehicle with this license plate already exists");
            }
            vehicle.setLicensePlate(dto.getLicensePlate());
        }

        if (dto.getWeightLimit() != null) {
            vehicle.setWeightLimit(dto.getWeightLimit());
        }

        if (dto.getBrandId() != null) {
            VehicleBrand brand = brandRepository.findById(dto.getBrandId())
                    .orElseThrow(() -> new RuntimeException("Brand not found"));
            vehicle.setBrand(brand);
        }

        if (dto.getModelId() != null) {
            VehicleModel model = modelRepository.findById(dto.getModelId())
                    .orElseThrow(() -> new RuntimeException("Model not found"));

            if (!model.getBrand().getId().equals(vehicle.getBrand().getId())) {
                throw new RuntimeException("Model does not belong to the selected brand");
            }
            vehicle.setModel(model);
        }

        if (dto.getImagesToDelete() != null && !dto.getImagesToDelete().isEmpty()) {
            List<VehicleImage> imagesToRemove = vehicle.getImages().stream()
                    .filter(img -> dto.getImagesToDelete().contains(img.getId()))
                    .toList();

            for (VehicleImage img : imagesToRemove) {
                minioService.deleteFile(img.getMinioBucket(), img.getMinioPath());
                vehicle.getImages().remove(img);
            }
        }

        if (newImages != null && !newImages.isEmpty()) {
            for (MultipartFile file : newImages) {
                String minioPath = minioService.uploadFile(file, vehicleImagesBucket, "vehicle-" + vehicle.getId());

                VehicleImage img = new VehicleImage();
                img.setOriginalName(file.getOriginalFilename());
                img.setMinioPath(minioPath);
                img.setMinioBucket(vehicleImagesBucket);
                img.setVehicle(vehicle);
                vehicle.getImages().add(img);
            }
        }

        vehicle.setUpdatedAt(LocalDateTime.now());
        vehicle = vehicleRepository.save(vehicle);

        return VehicleResponseDTO.fromEntity(vehicle);
    }

    @Transactional
    @Caching(evict = {
        @CacheEvict(value = "vehicleById", key = "#id"),
        @CacheEvict(value = "vehiclesPage", allEntries = true),
        @CacheEvict(value = "vehicleSearch", allEntries = true)
    })
    public void delete(Long id) {
        Vehicle vehicle = vehicleRepository.findByIdWithDetails(id)
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));

        for (VehicleImage img : vehicle.getImages()) {
            minioService.deleteFile(img.getMinioBucket(), img.getMinioPath());
        }

        vehicleRepository.delete(vehicle);
    }

    @Cacheable(value = "vehicleBrands")
    public List<VehicleBrand> getAllBrands() {
        return brandRepository.findAll();
    }

    @Cacheable(value = "vehicleModels", key = "#brandId")
    public List<VehicleModel> getModelsByBrand(Long brandId) {
        return modelRepository.findByBrandId(brandId);
    }

    @Cacheable(value = "vehicleLocation", key = "#vehicleId")
    public VehicleLocationDTO getLastLocation(Long vehicleId) {
        Vehicle vehicle = vehicleRepository.findById(vehicleId)
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));

        return locationRepository.findByVehicleId(vehicleId)
                .map(VehicleLocationDTO::fromEntity)
                .orElse(null);
    }

    @Cacheable(value = "distanceStats", key = "#vehicleId + '-' + #startDate + '-' + #endDate")
    public DistanceStatisticsDTO getDistanceStatistics(Long vehicleId, LocalDate startDate, LocalDate endDate) {
        Vehicle vehicle = vehicleRepository.findById(vehicleId)
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));

        return telemetryService.getAggregatedDistance(vehicleId, vehicle.getLicensePlate(), startDate, endDate);
    }

    @Cacheable(value = "availabilityStats", key = "#vehicleId + '-' + #startTime.toEpochMilli() + '-' + #endTime.toEpochMilli()")
    public AvailabilityStatisticsDTO getAvailabilityStatistics(Long vehicleId, Instant startTime, Instant endTime) {
        Vehicle vehicle = vehicleRepository.findById(vehicleId)
                .orElseThrow(() -> new RuntimeException("Vehicle not found"));

        return telemetryService.getAggregatedAvailability(vehicleId, vehicle.getLicensePlate(), startTime, endTime);
    }
}